docs/design/AnoncredsW3CCompatibility.md

```diff
@@ -0,0 +1,286 @@
+
+### What are the functions we are going to wrap?
+
+After thoroughly reviewing upcoming changes from [anoncreds-rs PR273](https://github.com/hyperledger/anoncreds-rs/pull/273), the classes or `AnoncredsObject` that we deemed necessary to be exported are:<br>
+
+[W3CCredentialOffer](https://github.com/hyperledger/anoncreds-rs/pull/273/files#diff-6f8cbd34bbd373240b6af81f159177023c05b074b63c7757fc6b3796a66ee240R106)<br>
+class methods (`create`, `load`)<br>
+bindings functions (`create_w3c_credential_offer`)<br>
+
+[W3CCredentialRequest](https://github.com/hyperledger/anoncreds-rs/pull/273/files#diff-6f8cbd34bbd373240b6af81f159177023c05b074b63c7757fc6b3796a66ee240R162)<br>
+class methods (`create`, `load`)<br>
+bindings functions (`create_w3c_credential_request`)<br>
+
+[W3CCredential](https://github.com/hyperledger/anoncreds-rs/pull/273/files#diff-6f8cbd34bbd373240b6af81f159177023c05b074b63c7757fc6b3796a66ee240R424)<br>
+class methods (`create`, `load`)<br>
+instance methods (`proceess`, `to_legacy`, `add_non_anoncreds_integrity_proof`, `set_id`, `set_subject_id`, `add_context`, `add_type`)<br>
+class properties (`schema_id`, `cred_def_id`, `rev_reg_id`, `rev_reg_index`)<br>
+bindings functions (`create_w3c_credential`, `process_w3c_credential`, `_object_from_json`, `_object_get_attribute`, `w3c_credential_add_non_anoncreds_integrity_proof`, `w3c_credential_set_id`, `w3c_credential_set_subject_id`, `w3c_credential_add_context`, `w3c_credential_add_type`)<br>
+
+[W3CPresentation](https://github.com/hyperledger/anoncreds-rs/pull/273/files#diff-6f8cbd34bbd373240b6af81f159177023c05b074b63c7757fc6b3796a66ee240R791)<br>
+class methods (`create`, `load`)<br>
+instance methods (`verify`)<br>
+bindings functions (`create_w3c_presentation`, `_object_from_json`, `verify_w3c_presentation`)<br>
+
+They will be added to [\_\_init\_\_.py](https://github.com/hyperledger/anoncreds-rs/blob/main/wrappers/python/anoncreds/__init__.py) as additional exports of AnoncredsObject. <br><br>
+
+We also have to consider which classes or anoncreds objects have been modified
+
+The classes modified according to the same [PR](https://github.com/hyperledger/anoncreds-rs/pull/273) mentioned above are:<br>
+
+[Credential](https://github.com/hyperledger/anoncreds-rs/pull/273/files#diff-6f8cbd34bbd373240b6af81f159177023c05b074b63c7757fc6b3796a66ee240R402)<br>
+added class methods (`from_w3c`)<br>
+added instance methods (`to_w3c`)<br>
+added bindings functions (`credential_from_w3c`, `credential_to_w3c`)<br>
+
+[PresentCredential](https://github.com/hyperledger/anoncreds-rs/pull/273/files#diff-6f8cbd34bbd373240b6af81f159177023c05b074b63c7757fc6b3796a66ee240R603)<br>
+modified instance methods (`_get_entry`, `add_attributes`, `add_predicates`)<br>
+<br>
+
+### How will they fit into ACA-PY?
+
+There are two scenarios to consider when we want to add w3c format support to ACA-PY.
+
+- Creating a W3C VC credential from credential definition, and issuing and presenting it as is
```

Comment:
These are the places that we need to focus -- how is the ACA-py Admin API affected to enable creating an AnonCreds VC and what happens after that.  I think we have to make the technique backwards compatible, so that existing Controllers (business logic) continue to work with legacy AnonCreds.

The options that I see are to add an optional flag "W3cFormat" (or whatever 😄 ) that is set true when an issuer wants to issue in W3C format.  Alternatively, we can implement different endpoints.

For receiving a credential issued in W3C format, or a presentation presented in W3C format, we'd like that to "just work". Detect the format and process the data in either format.  The tricky thing there is going to be the (still to be determined) attachment format of an AnonCreds credential/presentation.  That topic is still being discussed.

Reply:
I suggest using a flag named `VCDM11` and another named `VCVDM20`, since this reduces the differences between them to a trivial constant, and sets up ACA-Py for future adaptations if the VCDM evolves more.

Reply:
I spent some time poking around the admin interface and studying test code, and I'm thinking that @swcurran is a genius. My vision wasn't good enough. Stephen said that he prefers to have ACA-Py just do the right thing when it is receiving a credential. I was thinking that on the issuance side, we'd have to add a flag to `create_credential` (e.g., [here](https://github.com/hyperledger/aries-cloudagent-python/blob/6c7827e63303d76c940a903e656e0e238fdc1926/aries_cloudagent/anoncreds/issuer.py#L546)), and so I was looking at higher-level interfaces to try to figure out how a caller could provide input in a way that would tell us how to set that flag. But now I'm thinking that it is also possible to extend the "just do the right thing, based on the data" strategy to issuance as well. This would eliminate the need to change the admin interface at all, because the data format would be signaled by the credential request. We'd still modify `create_credential`, maybe, but only to accept a flag and do a credential conversion at the very end; otherwise everything would stay the same.

The magic is enabled by using the data payload of the [`propose-credential` message in the Issue Credential protocol](https://github.com/hyperledger/aries-rfcs/blob/main/features/0453-issue-credential-v2/README.md#propose-credential). If the preview is a w3c format, the everything switches to issuing in that format. If not, it remains legacy. This means that ACA-Py just always, naturally and perfectly, does the right thing, serving up whatever data type the remote parties want.

Maybe there are some gotchas to this approach that I haven't realized, but it seems ideal if we can make it work.

Reply:
yeah i like this approach - is the old Perl spirit, and simpler semantics